aten/src/ATen/Dispatch.h

```diff
@@ -1,65 +1,39 @@
-    case ScalarType::Char:
-      return F<int8_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Double:
-      return F<double>::apply(std::forward<Args>(args)...);
-    case ScalarType::Float:
-      return F<float>::apply(std::forward<Args>(args)...);
-    case ScalarType::Int:
-      return F<int>::apply(std::forward<Args>(args)...);
-    case ScalarType::Long:
-      return F<int64_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Short:
-      return F<int16_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Half:
-      return F<Half>::apply(std::forward<Args>(args)...);
-    default:
-      runtime_error("%s not implemented for '%s'", name, the_type.toString());
-    }
-}
-
-template<typename R, template <typename> class F, typename ... Args>
-R dispatch_floating_types(const Type& the_type, const char *name, Args&&... args) {
-  switch(the_type.scalarType()) {
-    case ScalarType::Double:
-      return F<double>::apply(std::forward<Args>(args)...);
-    case ScalarType::Float:
-      return F<float>::apply(std::forward<Args>(args)...);
-    case ScalarType::Half: // no native half math on either CPU or GPU.
-    default:
-      runtime_error("%s not implemented for '%s'", name, the_type.toString());
+#define AT_PRIVATE_CASE_TYPE(enum_type, type, function) \
+  case enum_type: {                                     \
+    using scalar_t = type;                              \
+    return function();                                  \
   }
-}
 
-#define DISPATCH_ALL_FLOATING_TYPES(TYPE, NAME, ...)                             \
-  [&]() {                                                                         \
-    const Type& the_type = TYPE;                                                  \
-    switch (the_type.scalarType()) {                                              \
-      case ScalarType::Double: {                                                  \
-        using scalar_t = double;                                                  \
-        return __VA_ARGS__();                                                               \
-      } case ScalarType::Float: {                                                 \
-        using scalar_t = float;                                                   \
-        return __VA_ARGS__();                                                               \
-      } case ScalarType::Half: {                                                  \
-        using scalar_t = half;                                                    \
-        return __VA_ARGS__();                                                               \
-      } default:                                                                  \
-        runtime_error("%s not implemented for '%s'", NAME, the_type.toString());  \
-    }                                                                             \
+#define AT_DISPATCH_FLOATING_TYPES(TYPE, NAME, function)                 \
```

Comment:
I liked the `...`/`__VA_ARGS__` trick because macros have stupid parsing rules and don't handle lambdas well. If you forget the `()` around the lambda expression you get hard to decipher error messages.

Reply:
Yeah I see your point. I think when @zdevito and I looked at it we were a bit perplexed by the need for `__VA_ARGS__` as it was a bit unintuitive. It's also [not standard](https://stackoverflow.com/questions/5588855/standard-alternative-to-gccs-va-args-trick) btw, so it'll eventually fail on some compiler...
Like, what happens when you pass an argument between the `NAME` and the lambda by accident? It would get accepted right now and fail in super weird ways inside the macro

Reply:
It's standard C++11 (and C99): http://en.cppreference.com/w/cpp/preprocessor/replace

Reply:
Oh ok, my bad. Both approaches seem to have drawbacks. But I can change it back to `__VA_ARGS__` then